File: src/main/java/net/effize/bandlog/team/application/TeamCommandService.java
package net.effize.bandlog.team.application;

import net.effize.bandlog.team.application.dto.*;
import net.effize.bandlog.team.domain.model.Member;
import net.effize.bandlog.team.domain.model.MemberRole;
import net.effize.bandlog.team.domain.model.Team;
import net.effize.bandlog.team.domain.model.TeamId;
import net.effize.bandlog.team.domain.service.TeamDomainService;
import net.effize.bandlog.team.domain.model.UserId;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.List;

@Service
@Transactional
public class TeamCommandService {
    private final TeamDomainService teamDomainService;
    private final TeamRepositoryService teamRepositoryService;

    public TeamCommandService(TeamDomainService teamDomainService, TeamRepositoryService teamRepositoryService) {
        this.teamDomainService = teamDomainService;
        this.teamRepositoryService = teamRepositoryService;
    }

    public CreateTeamResponse createTeam(UserId authUserId, CreateTeamRequest request) {
        Instant now = Instant.now();
        
        // Domain Service로 비즈니스 로직 처리
        Team newTeam = teamDomainService.createTeam(request.name(), request.description(), now);
        Team savedTeam = teamRepositoryService.saveTeam(newTeam);
        
        // 팀 생성자를 리더로 추가
        Member leaderMember = teamDomainService.addMemberToTeam(savedTeam, authUserId, MemberRole.LEADER, now);
        teamRepositoryService.saveMember(leaderMember);

        return new CreateTeamResponse(savedTeam.id().value());
    }

    public RefreshTeamInviteCodeResponse refreshTeamInviteCode(UserId authUserId, RefreshTeamInviteCodeRequest request) {
        Team foundTeam = teamRepositoryService.findTeamById(TeamId.of(request.teamId()));
        List<Member> members = teamRepositoryService.findMembersByTeam(foundTeam);
        
        // Domain Service로 권한 검증
        teamDomainService.validateTeamLeaderPermission(members, authUserId);

        foundTeam.refreshInviteCode();
        teamRepositoryService.saveTeam(foundTeam);
        
        return new RefreshTeamInviteCodeResponse(foundTeam.id().value());
    }

    public JoinTeamResponse joinTeam(UserId authUserId, JoinTeamRequest request) {
        Instant now = Instant.now();
        Team foundTeam = teamRepositoryService.findTeamById(TeamId.of(request.teamId()));
        List<Member> members = teamRepositoryService.findMembersByTeam(foundTeam);
        
        // Domain Service로 비즈니스 규칙 검증
        teamDomainService.validateUserNotAlreadyMember(members, authUserId);

        Member newMember = teamDomainService.addMemberToTeam(foundTeam, authUserId, MemberRole.MEMBER, now);
        teamRepositoryService.saveMember(newMember);

        return new JoinTeamResponse(foundTeam.id().value());
    }
}

File: src/main/kotlin/net/effize/bandlog/team/application/TeamCommandService.kt
package net.effize.bandlog.team.application

import net.effize.bandlog.common.id.UserId
import net.effize.bandlog.team.adapter.`in`.web.request.CreateTeamRequest
import net.effize.bandlog.team.adapter.`in`.web.request.JoinTeamRequest
import net.effize.bandlog.team.adapter.`in`.web.request.RefreshTeamInviteCodeRequest
import net.effize.bandlog.team.adapter.`in`.web.response.CreateTeamResponse
import net.effize.bandlog.team.adapter.`in`.web.response.JoinTeamResponse
import net.effize.bandlog.team.adapter.`in`.web.response.RefreshTeamInviteCodeResponse
import net.effize.bandlog.team.domain.MemberRole
import net.effize.bandlog.team.domain.TeamId
import org.springframework.stereotype.Service
import java.time.Instant

@Service
class TeamCommandService(
    private val teamService: TeamService
) {

    fun createTeam(authUserId: UserId, request: CreateTeamRequest): CreateTeamResponse {
        val now = Instant.now()
        val createdTeam = teamService.createTeam(request.name, request.description, now)
        teamService.addNewMember(createdTeam, authUserId, MemberRole.LEADER, now)

        return CreateTeamResponse(createdTeam.id().value)
    }

    fun refreshTeamInviteCode(authUserId: UserId, request: RefreshTeamInviteCodeRequest): RefreshTeamInviteCodeResponse {
        val foundTeam = teamService.activeTeam(TeamId.of(request.teamId))
        val members = teamService.membersOf(foundTeam)
        val meLeaderCount = members
            .filter { it.userId() == authUserId }
            .count { it.role == MemberRole.LEADER }

        if (meLeaderCount <= 0) throw IllegalStateException("User is not a leader of the team")

        foundTeam.refreshInviteCode()
        return RefreshTeamInviteCodeResponse(foundTeam.id().value)
    }

    fun joinTeam(authUserId: UserId, request: JoinTeamRequest): JoinTeamResponse {
        val now = Instant.now()
        val foundTeam = teamService.activeTeam(TeamId.of(request.teamId))
        val members = teamService.membersOf(foundTeam)
        val meCount = members.count { it.userId() == authUserId }

        if (meCount > 0) throw IllegalStateException("User already joined the team")

        teamService.addNewMember(foundTeam, authUserId, MemberRole.MEMBER, now)

        return JoinTeamResponse(foundTeam.id().value)
    }
}
